x/ibc-rate-limit/ics4_wrapper.go

```diff
@@ -84,6 +91,12 @@ func (i *ICS4Wrapper) GetParams(ctx sdk.Context) (contract string) {
 
 // CalculateChannelValue The value of an IBC channel. This is calculated using the denom supplied by the sender.
 // if the denom is not correct, the transfer should fail somewhere else on the call chain
-func (i *ICS4Wrapper) CalculateChannelValue(ctx sdk.Context, denom string) sdk.Int {
-	return i.bankKeeper.GetSupplyWithOffset(ctx, denom).Amount
+func (i *ICS4Wrapper) CalculateChannelValue(ctx sdk.Context, denom string, packet exported.PacketI) sdk.Int {
+	// If the source is the counterparty chain, this should be
+	if strings.HasPrefix(denom, "ibc/") {
+		return i.bankKeeper.GetSupplyWithOffset(ctx, denom).Amount
+	}
+
+	escrowAddress := transfertypes.GetEscrowAddress(packet.GetSourcePort(), packet.GetSourceChannel())
+	return i.bankKeeper.GetBalance(ctx, escrowAddress, denom).Amount
```

Comment:
At first look this doesn't seem correct. If we have something like:
```
 A            B              C
ch0  --- ch1    ch2 --- ch3
```
And `a` is chain A token. In the following cases:
1. B sends to C a packet with denom `ch1/a` over `ch2`. denom here is `ch1/a`. 
2. B receives (back) from C a packet with denom `ch3/ch1/a` over `ch2`. denom here is `ch1/a` (unprefixed)

Here I think we get the `ch1/a` supply instead of the balance of `escrow_address(ch2)` for denom `ch1/a`.

Maybe something along these lines:
```suggestion
func (i *ICS4Wrapper) CalculateChannelValue(ctx sdk.Context, denom string, packet exported.PacketI, send bool) sdk.Int {
	prefix_for_source:= transfertypes.GetDenomPrefix(packet.GetSourcePort(), packet.GetSourceChannel())
	if send { // send packet
		if !strings.HasPrefix(denom, prefix_for_source) {
			// Sender chain is the source
			escrowAddress := transfertypes.GetEscrowAddress(packet.GetSourcePort(), packet.GetSourceChannel())
			return i.bankKeeper.GetBalance(ctx, escrowAddress, denom).Amount
		} else {
			return i.bankKeeper.GetSupplyWithOffset(ctx, denom).Amount
		}
	} else { // receive packet
		if strings.HasPrefix(denom, prefix_for_source) {
			// Receiver chain is the source
			escrowAddress := transfertypes.GetEscrowAddress(packet.GetDestPort(), packet.GetDestChannel())
			return i.bankKeeper.GetBalance(ctx, escrowAddress, denom).Amount
		} else {
			return i.bankKeeper.GetSupplyWithOffset(ctx, denom).Amount
		}
	}
```

It would be great to have tests to cover scenarios like above.

Reply:
The current behaviour (what you described above) is what I was trying to achieve, where escrow value is *only* used for native tokens (osmo/ion/tokenfactory tokens). ch0/a, as an ibc token, should be rate limited based on the current  amount of ch0/a on osmosis. 

I'm open to changing it to what you suggest, but I think the volume of some token/channel combinations may be too low and make the limit to be hit too often

Reply:
> The current behaviour (what you described above) is what I was trying to achieve, where escrow value is _only_ used for native tokens (osmo/ion/tokenfactory tokens). ch0/a, as an ibc token, should be rate limited based on the current amount of ch0/a on osmosis.
> 
I see, thanks for the clarification.

> I'm open to changing it to what you suggest, but I think the volume of some token/channel combinations may be too low and make the limit to be hit too often

But in this case ratelimiting should not be configured on those channels as they have a low value.
Are there osmosis uscases like the ones I mention above (the multi hop/channel transfers)?